scripts/rust-libc.py

```diff
@@ -0,0 +1,581 @@
+#!/bin/env python3
+
+# HOW THIS WORKS
+#
+# This script takes mlibc header files and generates bindings to be used with rust's "libc" crate.
+# A configuration file is needed for its proper function; an example is provided alongside this
+# script. Please do note that it is used for managarm, which lives under `unix/linux_like` in the
+# "libc" crate. If your OS does not live under this directory, but e.g. under just `unix` instead,
+# you will need to adapt the configuration to fit your use.
+#
+# HOW TO USE
+#
+# > python rust-libc <path/to/your/installed/mlibc/headers> [<single-header.h>]
+#
+# By default, the script parses all header files in the directory supplied, except for when a
+# single header is provided, where it will only parse that.
+
+import argparse
+import binascii
+import clang.cindex
+import colorama
+import functools
+import io
+import itertools
+import pathlib
+import string
+import sys
+import yaml
+
+from clang.cindex import (TypeKind, CursorKind, TokenKind, Cursor)
+
+dry_run = True
+errors_emitted = 0
+
+def log_err(prefix, msg):
+	global errors_emitted
+
+	print(f"{colorama.Fore.RED}{prefix}{colorama.Style.RESET_ALL}: {msg}", file=sys.stderr)
+	errors_emitted += 1
+
+def emit(msg):
+	if not dry_run:
+		print(msg)
+
+def no_system_includes(cursor, level):
+	'''filter out verbose stuff from system include files'''
+	return (level!= 1) or (cursor.location.file is not None and not cursor.location.file.name.startswith('/usr/include'))
+
+class Type:
+	def __init__(self, c: clang.cindex.Cursor, t: clang.cindex.Type = None, convert_arrays_to_ptrs = False):
+		self.cursor = c
+		self.type = t if t else c.type
+		self.convert_arrays_to_ptrs = convert_arrays_to_ptrs
+
+	@property
+	def kind(self):
+		return self.type.kind
+
+	def convert_ptr_type(self, c, ty, is_pointee=False):
+		pointee = ty if is_pointee else ty.get_pointee()
+
+		if pointee.kind == TypeKind.FUNCTIONPROTO:
+			arg_list = []
+			for f in pointee.argument_types():
+				arg_list.append(f"{Type(c, f)}")
+			args = ", ".join(arg_list)
+			ret_type = Type(c, pointee.get_result())
+			if c.semantic_parent.spelling in config["force_raw_function_pointer"]:
+				return f"extern \"C\" fn({args})" + (f" -> {ret_type}" if ret_type != "::c_void" else "")
+			else:
+				return f"::Option<unsafe extern \"C\" fn({args})" + (f" -> {ret_type}>" if ret_type != "::c_void" else ">")
+
+		is_mut = not pointee.spelling.startswith("const")
+		prefix = "*" + ("mut" if is_mut else "const") + " "
+
+		type_iter = pointee
+
+		while type_iter.kind == TypeKind.POINTER:
+			prefix += "*mut "
+			type_iter = type_iter.get_pointee()
+
+		t = type_iter.spelling.removeprefix("const ")
+		tokens = t.split(' ')
+
+		match tokens[0]:
```

Comment:
you can match multiple values at once, nesting matches is nearly always wrong

Reply:
actually, even better, this is an array

```py
match tokens:
    case ["char", *_]:
        return f"{prefix}::c_char"
    case ["struct", x, *_] if x in config["force_local_types"]:
        return prefix + x
    case ["struct", x, *_]:
        return f"{prefix}::{x}"
    case ["int", *_]:
...
```

note that many of the `*_` are there only because your original code doesn't length-check, so I eagerly ignore other entries.  if you know sizes in advance, you don't need to do that:

```py
>>> def f(x):  # bad formatting, repl code
...     match x:
...             case ["foo", *_]:
...                     print("foo")
...             case ["bar"]:
...                     print("bar")
...             case _:
...                     print("gone")
... 
>>> f(["foo", 1, 2, 3])
foo
>>> f(["foo"])
foo
>>> f(["bar"])
bar
>>> f(["bar", 1])
gone
>>> 
```

Reply:
Applied.